lib/routemap.c

```diff
@@ -1755,7 +1786,14 @@ static int route_map_dep_update(struct hash *dephash, const char *dep_name,
 		if (!dep->this_hash)
 			dep->this_hash = dephash;
 
-		hash_get(dep->dep_rmap_hash, rname, route_map_name_hash_alloc);
+		memset(&tmp_dep_data, 0, sizeof(struct route_map_dep_data));
```

Comment:
So with this change if I do this:

route-map FOO permit 10
   match mac address 01:02:03:04:05:06
   match mac address 01:02:03:04:05:06
   match mac address 01:02:03:04:05:06

the dep_data->refcnt is now 3 and we only have 1 reference.    We need additional commits to go through every possible code path where this function is called and make sure that duplicate config is safely `ignored`.

Additionally lines 1765 and 1766 while you did not touch them, clearly look like we have created a memory leak.

Reply:
Thanks, @donaldsharp for pointing this out. I can add the changes to handle the duplicate configurations as part of this PR itself.
The PR https://github.com/FRRouting/frr/pull/4287 already handles the duplicate configurations. However, since the return code is RMAP_COMPILE_SUCCESS, the dependents are once again updated which results in the reference counter to be incremented incorrectly.
When a duplicate configuration is encountered, we need to probably return something like RMAP_DUPLICATE_RULE and prevent the code from updating the dependency reference counter.

In addition to this, we need to handle the case below.

```route-map FOO permit 10
match mac address 00:00:00:00:00:0a
match mac address 00:00:00:00:00:0b```

This is a case of rule replacement. When "00:00:00:00:00:0a" is configured the reference counter for "00:00:00:00:00:0a" is set to 1.
When it is replaced by "00:00:00:00:00:0b", the reference counter for "00:00:00:00:00:0a" should be decremented by 1.
Let me try to handle this in this PR as well.

Finally, about the potential memory leak, the code does not return abruptly. It uses the goto statement ```goto out```
The memory allocated in 1765 and 1766 is freed up here.